GGRC-4179,4264,4415 Fix creation of Audit stub in Assessment Template model

neocoder opened the pull request:
This PR solves 3 issues as they are all connected.

# Dependency
- [x] GGRC-4355 (#7243) Back-end
- [x] #7298 Issue Tacker mock, to fully test changes in this PR

# Issues descriptions

__issue 1__
When Assessment Template form is created without link to the Audit, the audit stub is created
for in the form_preload hook of the AssessmentTemplate model. Later in after_save hook, reify() and refresh() method are called on the stub. The issue is that audit stub was created as plain js object and did not have those methods.

_Steps to reproduce:_
1. Go to the audit page
2. Create Assessment template
3. Hover over Assessment template first tier and open Edit Assessment template modal window
4. Click Save button w/o any changes

Prior to the changes in this PR and error was thrown in the console: "Uncaught TypeError: this.audit.reify is not a function".

__issue 2__
An error is thrown in the console that `issue_tracker` property is undefined when user tries to edit Assessment Template from Global Search

_Steps to reproduce:_
1. Log as Admin user and open My work page
2. Click Global Search button and select Assessment Template object type > click Search
3. Click grey triangle in Search results
4. Click Edit Assessment template button in 3 bb's menu

Prior to the changes in this PR an error "Uncaught TypeError: Cannot read property 'issue_tracker' of undefined" error was displayed in the console

__issue 3__
Cannot switch on or off Issue Tracker integration for assessments that were created when Issue Tracker was switched off for audit.

_Steps to reproduce:_
1. Have an audit where bugunazer integration is switched off
2. Create an Assessment
3. Switch on Issue Tracker for audit
4. Invoke Edit Assessment popup and try to switch switch on Issue Tracker integration

Prior to the changes in this PR user was not able to see Issue Tracker controls even if Issue Tracker was enabled for Audit.

__issue 4__
`Cannot read property 'issue_tracker' of undefined"` error occurs when clicking the  "Define Assessment Template" link 

_Steps to reproduce:_
1. Open My work page > Audit tab
2. Expand Audit Info panel 
3. Click Define Assessment template link in 3 bb's menu

Prior to the changes in this PR and error "Uncaught (in promise) TypeError: Cannot read property 'issue_tracker' of undefined" appeared in the console.

# Solution description

__issue 1__
Remove `after_save` event handle in Assessment model as the case for which this code was added is no longer valid.

__issue 2__
The error happed because there's no audit on the assessment template searched from the Global Search. The solution is to fetch parent Audit from server if Assessment Template is edited from Global Search.

__issue 3__
Refactor work with Issue Tracker as there are some logical errors in the code.

__issue 4__
Pass issue_tracker property in data-object-params when creating the "Define Assessment Template" modal

# Sanity checklist

- [x] I have clicked through the app to make sure my changes work and not break the app.
- [x] I have applied the correct milestone and labels.
- [x] My changes fix the issue described in the description (and do nothing else). 🤞
- [x] My changes are covered by tests.
- [x] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [x] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [x] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).

# PR Review checklist

- [ ] The changes fix the issue and don't cause any apparent regressions.
- [ ] Labels and milestone are correctly set.
- [ ] The solution description matches the changes in the code.
- [ ] There is no apparent way to improve the performance & design of the new code.
- [ ] The pull request is opened against the correct base branch.
- [ ] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

AlexNo commented:
`code style` failed

AlexNo commented on src/ggrc/assets/javascripts/models/audit_models.js:
```diff
@@ -301,13 +301,13 @@ import Permission from '../permission';
           context: pageInstance.context,
           issue_tracker: pageInstance.issue_tracker,
-        };
+        }).stub();
```
After call `stub()` method on instance we will lose `issue_tracker` , `context ` and `title` fields, which required for correct work issue tracker integration. You can clarify this point with @YuriChupyrkin

neocoder commented:
@AlexNo I removed the stub() as we discussed.
Also, I added the `on hold` label because I found an issue with Issue Tracker integration and need to figure out whether we had a ticket for it.

zidarsk8 commented on src/ggrc-client/js/models/assessment.js:
```diff
@@ -315,6 +320,20 @@ import {prepareCustomAttributes} from '../plugins/utils/ca-utils';
+          id: this.id,
+        }, ['id', 'title', 'type', 'context', 'issue_tracker']);
+
+        makeRequest({data: [param]}).then((response)=> {
```
On what pages is this used and why is the audit not available on those?

This seems like a hack and a performance degradation, so if this truly is needed, then the else block needs a good comment explaining why this request is  performed here and in which cases this can happen.

neocoder replied:
@zidarsk8 thanks for the comment. I didn't update the PR description yet, as I'm waiting for another ticket, which this PR fixes, to be created. This particular call is needed when you are opening Global Search outside of Audit page, search for Assessement Template and open its edit modal from there. In this case we need to fetch Audit to get issue_tracker options. I don't think this will be a performance degradation as this scenario is not commonly used(and doesn't work in current implementation at all. see ticket GGRC-4179). I will add a comment block above that request call.

zidarsk8 replied:
@neocoder Thanks for explanation. And yes a comment with this above the request will be useful.

neocoder commented:
Setting `needs work` for now. Need to check for performance regressions.

AleksNeStu commented:
@neocoder I rerun selenium to due to the previous one wasn't available.

AleksNeStu commented:
@neocoder selenium tests got an issue in a moment of trying create new asmts by pressing create btn on asmts' tab tree view under audit.

![image](https://user-images.githubusercontent.com/16610768/35925449-086957d4-0c26-11e8-8b6c-155b93f34d16.png)

NastassiaKiseleva commented:
@YuriChupyrkin please take over here, please finish this PR - unit tests are failed now, adding needs work

AleksNeStu commented:
@NastassiaKiseleva @YuriChupyrkin I deleted 'Needs selenium fix'  due to mentioned above regression issue wasn't fixed and for now we don't need any efforts to fix.

YuriChupyrkin commented:
Changes from this PR moves to #7313.

This PR can be closed.

AlexNo commented:
Closed in favor https://github.com/google/ggrc-core/pull/7313
